File: practice/algorithms/strings/game_of_thrones_1/GameOfThrones1.java
import java.util.*;

public class GameOfThrones1 {

    public static void main(String[] args) {
        Scanner stdin = new Scanner(System.in);
        String line = stdin.nextLine();
        Map<Character, Integer> charMap = new HashMap<>();
        for(int i = 0; i < line.length(); i++) {
            char c = line.charAt(i);
            int count = charMap.getOrDefault(c, 0);
            charMap.put(c, count + 1);
        }
        System.out.println(isPalindrome(line.length(), charMap));
        stdin.close();
    }

    private static String isPalindrome(int length, Map<Character, Integer> charMap) {
        boolean isPalindrome = true;
        if(length % 2 == 0) {
            for(char key : charMap.keySet()) {
                if(charMap.get(key) % 2 != 0) {
                    isPalindrome = false;
                    break;
                }
            }
        }
        if(length % 2 != 0) {
            int oddCount = 0;
            for(char key : charMap.keySet()) {
                if(charMap.get(key) % 2 != 0) {
                    oddCount++;
                }
            }
            isPalindrome = oddCount == 1;
        }
        if(isPalindrome) {
            return "YES";
        } else {
            return "NO";
        }
    }
}

File: practice/algorithms/strings/game_of_thrones_1/GameOfThrones1.scala
import scala.io.Source

object GameOfThrones1 extends App {
  val line = Source.stdin.bufferedReader().readLine()
  validate(line)

  def isPalindrome(length: Int, charMap: Map[Char, String]): Boolean = {
    if(length % 2 == 0)
      charMap.forall(entry => entry._2.length % 2 == 0)
    else
      charMap.count(entry => entry._2.length % 2 == 1) == 1
  }

  def validate(line: String): Unit = {
    val charMap = line.groupBy(c => c)
    println(if(isPalindrome(line.length, charMap)) "YES" else "NO")
  }
}